### PR TITLE
Add default impl. to RootAction

### PR DESCRIPTION
**Pull request by Wadeck**
- to ease the inner implementation that often return null

Proposed after discussion in JTH.

If this one is accepted, we can close the [#85 JTH](https://github.com/jenkinsci/jenkins-test-harness/pull/85)

@reviewbybees

### CODE REVIEW

**Comment by daniel-beck** [comment]
Why aren't we doing that with `Action` to begin with? Perhaps we want to make the overriding explicit, but then, why is this proposing to do that with `RootAction`?

**Comment by Wadeck** [comment]
@daniel-beck I changed accordingly to your comment. I am not sure if I need to push the getDisplayName default method even to the `ModelObject` (parent of `Action`) since that interface does not provide `@CheckForNull` signature.

**Review by dwnusbaum — DISMISSED** [review]
🐝 I think this is preferable to the JTH PR. I'm not sure if pushing the default implementation of `getDisplayName` to `ModelObject` makes sense. The default is useful  for `Action`s since they are often invisible to the user, but I'm guessing (maybe incorrectly) that generally someone extending `ModelObject` wants to use a non-null display name.

**Comment by daniel-beck** [comment]
@jenkinsci/code-reviewers I need to know the exact shade of red for this bike shed 😉 

Is there value in continuing to force people to override the two (three?) methods of Action so they're not overlooked?

**Review by oleg-nenashev — COMMENTED** [review]
No strong opinion. It simplifies the test code, but there is unlikely a case for the production one. 

* In production we do expect people to implement these methods or to use https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/java/hudson/model/InvisibleAction.java.
* In tests: There is InvisibleAction.java for this case, so the tests can be simplified just by using this class.

Weak -1 regarding the change.

**Comment by jglick** [comment]
> There is `InvisibleAction` for this case, so the tests can be simplified just by using this class.

Ah but all methods in `InvisibleAction` are `final`, so you cannot use it as a base class if you do actually need to implement `getUrlName`, which is fairly common.

As I wrote in https://github.com/jenkinsci/jenkins-test-harness/pull/85 and @daniel-beck echoes, I am wary of people forgetting to override methods they actually need.

Making `InvisibleAction.getUrlName` non`final` would I think be the least controversial change, and satisfy most of the use cases. For that matter, doing nothing is OK too.

**Comment by daniel-beck** [comment]
> least controversial

I disagree. A visible action inheriting from `InvisibleAction` makes no sense.

**Comment by stephenc** [comment]
>  A visible action inheriting from `InvisibleAction` makes no sense.

But it would be invisible as it does not have an icon or display name so there would be no *visible* link.

Now if it was called `NonNavigableAction` you would have a point, but invisible != non-navigable, just we have chosen to make `InvisibleAction` really a `InvisibleNonNavigableAction`

**Comment by stephenc** [comment]
>  I am wary of people forgetting to override methods they actually need.

I am also wary... hence 👎  from my PoV to current change

**Comment by daniel-beck** [comment]
@stephenc 

> But it would be invisible as it does not have an icon or display name so there would be no *visible* link.

Good point, I missed @jglick specifically referring to `getUrlName`, and this PR implements the other two. So for purposes of `getDisplayName`, `getIconFileName` this still seems unsuitable.

**Comment by oleg-nenashev** [comment]
Yes, forgetting implementations was also my concern in the comment above. I'd guess there is an agreement that the change should be done in another ways. I second @stephenc and agree that a couple of new classes is feasible.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Requesting changes to follow the process

**Comment by oleg-nenashev** [comment]
@Wadeck ping, there is a pending feedback. Do you plan to work on this change?

**Comment by Wadeck** [comment]
@stephenc @oleg-nenashev I proposed a new approach based on a new class to not confuse user as discussed. I am not totally satisfied with the name `DefaultRootAction`, since the default comes from the default methods in the interface and not to be meant as "default rootAction". So if you have a good idea, I will take it.

(AbstractRootAction seems not really consistent with the use of an interface)

**Review by dwnusbaum — COMMENTED** [review]
I think rather than a new interface it would be preferable to make `InvisibleAction#getUrlName` non-final and have the test classes extend `InvisibleAction` and implement `RootAction` as discussed previously.

It seems like there aren't a ton of implementations that could be cleaned up by this change, so it might not be worth the work necessary to get everyone to agree. 😄

**Comment by jglick** [comment]
Urgh, [looks](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-3172/3/testReport/junit/org.jvnet.hudson.test/BuildTriggerTest/Windows___Windows_Publishing___testSomething/) like #3168 did not work.

File: core/src/main/java/hudson/model/DefaultRootAction.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * Simply {@link RootAction} with default implementation of method.
+ *
+ * <p>
+ * Extend from this interface and put {@link Extension} on your subtype
+ * to have them auto-registered to {@link jenkins.model.Jenkins}.
+ * 
+ * <p>
+ * Especially useful in test when you don't have to provide the methods returning null that
+ * does not provide any useful information for the tests.
+ *
```
Comment (jglick) [inline core/src/main/java/hudson/model/DefaultRootAction.java]: maybe `@see` from `InvisibleAction`

**Review by jglick — COMMENTED** [review]
I still feel like this is more noise than signal and it would be more intuitive to make `InvisibleAction.getUrlName` non`final` (and also useful for other kinds of `Action`s), but I do not have a strong opinion.

**Comment by jglick** [comment]
> did not work

Never mind, you just need to `git merge master` to bring this up to date.

**Comment by Wadeck** [comment]
Following advice from Devin + Jesse, `InvisibleAction#getUrlName` made final

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
IMHO it violates the `InvisibleAction` documentation: "Partial {@link Action} implementation that doesn't have any UI presence.". Apparently the refactored methods have such UI presence since you have to override the method

**Review by oleg-nenashev — APPROVED** [review]
weak 🐝

File: core/src/main/java/hudson/model/InvisibleAction.java [hunk 2]
```diff
@@ -24,12 +24,16 @@
 package hudson.model;
 
 /**
- * Partial {@link Action} implementation that doesn't have any UI presence.
+ * Partial {@link Action} implementation that doesn't have any UI presence (unless the {@link #getUrlName()} is overrided).
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/InvisibleAction.java]: This seems slightly confusing as just having a URL does not make for a UI presence, at least not in the most direct sense of having a presence on a side panel.

**Review by daniel-beck — APPROVED** [review]
:shipit:

File: core/src/main/java/hudson/model/InvisibleAction.java (hunk 2 above)
Comment (daniel-beck) [reply core/src/main/java/hudson/model/InvisibleAction.java]: In a way I'm contradicting what @oleg-nenashev wrote in his older comment. It really depends on what "UI presence" means. Perhaps just clarify that there can be a URL space with views etc., but there cannot be a presence in a sidepanel?

I guess unless you override `action.jelly` or what it's called. Well…

Let's just merge as is, any second spent on this is wasted.

**Comment by oleg-nenashev** [comment]
It was stalled for almost 6 months without new reviews. Taking the positive feedback in the latest comments, merging it

**Pull request merged**